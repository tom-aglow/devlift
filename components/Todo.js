import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, Text, View, Image, TouchableOpacity } from 'react-native';
import CheckBox from 'react-native-check-box';

class Row extends Component {
  handleToggleClick = () => {
    const { list, id, onToggle, isCompleted } = this.props;
    const newIsCompleted = !isCompleted;
    onToggle({ list, id, isCompleted: newIsCompleted });
  };

  handleDeleteClick = () => {
    const { id, onDelete, list } = this.props;
    onDelete({ list, id });
  };

  render() {
    const { text, isCompleted } = this.props;
    return (
      <View style={styles.container}>
        <CheckBox
          isChecked={isCompleted}
          onClick={this.handleToggleClick}
          checkedImage={
            <Image source={require('../img/checkbox_checked.png')} />
          }
          unCheckedImage={
            <Image source={require('../img/checkbox_unchecked.png')} />
          }
        />
        <View style={styles.textWrapper}>
          <Text style={[styles.text, isCompleted && styles.isCompleted]}>
            {text}
          </Text>
        </View>
        {isCompleted && (
          <TouchableOpacity onPress={this.handleDeleteClick}>
            <Image source={require('../img/button_delete.png')} />
          </TouchableOpacity>
        )}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    paddingVertical: 10,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between'
  },
  textWrapper: {
    flex: 1,
    marginLeft: 15
  },
  text: {
    fontSize: 16,
    fontFamily: 'OpenSans-Light'
  },
  isCompleted: {
    textDecorationLine: 'line-through'
  }
});

Row.propTypes = {
  list: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  text: PropTypes.string.isRequired,
  isCompleted: PropTypes.bool.isRequired,
  onToggle: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired
};

export default Row;
