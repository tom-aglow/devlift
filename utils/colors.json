{
  "white": "#FBFBFB",
  "lightGrey": "#E7E7E7",
  "mediumGrey": "#C2C2C2",
  "black": "#202020",
  "primary": "#12B4C2"
}
