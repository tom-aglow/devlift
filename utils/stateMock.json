{
  "todos": [
    { "id": 1, "text": "foo", "isCompleted": false, "listId": 0 },
    { "id": 2, "text": "bar", "isCompleted": true, "listId": 0 },
    { "id": 3, "text": "baz", "isCompleted": true, "listId": 1 }
  ],
  "lists": [
    { "id": 0, "title": "Personal" },
    { "id": 1, "title": "Movies to watch" }
  ],
  "inputValue": "",
  "editingId": 0,
  "openLists": [0]
}
