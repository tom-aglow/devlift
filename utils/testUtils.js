import React from "react";
import { mount, shallow } from "enzyme";
import { JSDOM } from "jsdom";

export const sel = id => `[data-test="${id}"]`;

export const setupGlobalObject = () => {
  const jsdom = new JSDOM("<!doctype html><html><body></body></html>");
  const { window } = jsdom;
  global.window = window;
  global.document = window.document;
};

export const TestHelper = class TestHelper {
  constructor(component, defaultProps) {
    this.component = component;
    this.defaultProps = defaultProps;
  }

  mountComponent(props = {}, isShallow = false) {
    const propsToUse = {
      ...this.defaultProps,
      ...props
    };

    if (isShallow) {
      return shallow(React.createElement(this.component, propsToUse));
    }

    return mount(React.createElement(this.component, propsToUse));
  }
};
