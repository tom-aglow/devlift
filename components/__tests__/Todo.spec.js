import Todo from '../Todo';
import { TestHelper, sel, setupGlobalObject } from '../../utils/testUtils';
import stateMock from '../../utils/stateMock.json';

const INDEX = 0;

const defaultProps = {
  ...stateMock.todos[INDEX],
  isEditing: false,
  onToggle() {},
  onEditToggle() {},
  onUpdate() {},
  onBlur() {},
  onDelete() {}
};

const helper = new TestHelper(Todo, defaultProps);

beforeAll(() => {
  setupGlobalObject();
});

it('renders correctly', () => {
  const wrapper = helper.mountComponent({}, true);
  expect(wrapper).toMatchSnapshot();
});

it('shows remove button if task is completed', () => {
  const wrapper = helper.mountComponent({ isCompleted: true }, true);
  const removeBtn = wrapper.find(sel('remove-todo-btn'));

  expect(removeBtn).toHaveLength(1);
});

it('does not show remove button if task is not completed', () => {
  const wrapper = helper.mountComponent({ isCompleted: false }, true);
  const removeBtn = wrapper.find(sel('remove-todo-btn'));

  expect(removeBtn).toHaveLength(0);
});

it('renders todo text when state is not editing', () => {
  const wrapper = helper.mountComponent({ isEditing: false }, true);
  const todoHtml = wrapper.find(sel('todo-text')).html();

  expect(todoHtml).toContain(stateMock.todos[INDEX].text);
});

it('renders input when state is editing', () => {
  const wrapper = helper.mountComponent({ isEditing: true }, true);
  const input = wrapper.find(sel('todo-edit-input'));

  expect(input).toHaveLength(1);
});

it('hides remove button in editing mode', () => {
  const wrapper = helper.mountComponent({ isEditing: true }, true);
  const removeBtn = wrapper.find(sel('remove-todo-btn'));

  expect(removeBtn).toHaveLength(0);
});

it('does not allow edit completed items', () => {
  const onEditToggle = jest.fn();
  const wrapper = helper.mountComponent(
    { isCompleted: true, onEditToggle },
    true
  );
  const todoText = wrapper.find(sel('todo-text'));
  todoText.props().onLongPress();

  expect(onEditToggle).toHaveBeenCalledTimes(0);
});
