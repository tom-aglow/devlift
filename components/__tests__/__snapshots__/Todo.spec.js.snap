// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "flex-start",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "paddingVertical": 5,
    }
  }
>
  <CheckBox
    checkedImage={
      <Image
        source={
          Object {
            "testUri": "../../../img/checkbox_checked.png",
          }
        }
        style={
          Object {
            "marginTop": 4,
          }
        }
      />
    }
    data-test="todo-checkbox"
    isChecked={false}
    isIndeterminate={false}
    leftTextStyle={Object {}}
    onClick={[Function]}
    rightTextStyle={Object {}}
    unCheckedImage={
      <Image
        source={
          Object {
            "testUri": "../../../img/checkbox_unchecked.png",
          }
        }
        style={
          Object {
            "marginTop": 4,
          }
        }
      />
    }
  />
  <TouchableOpacity
    activeOpacity={0.2}
    data-test="todo-text"
    onLongPress={[Function]}
    style={
      Object {
        "flex": 1,
        "marginLeft": 15,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontFamily": "OpenSans-Light",
            "fontSize": 16,
          },
          false,
        ]
      }
    >
      foo
    </Text>
  </TouchableOpacity>
</View>
`;
